mapo/agents/mapo/off_mapo_policy.py

```diff
@@ -71,20 +80,34 @@ def _build_critic_loss(policy, batch_tensors):
 
 def _build_actor_loss(policy, batch_tensors):
     obs = batch_tensors[SampleBatch.CUR_OBS]
-    policy_action = policy.model.get_actions(obs)
-    policy_action_value = policy.model.get_q_values(obs, policy_action)
-    policy.loss_stats["qpi_mean"] = tf.reduce_mean(policy_action_value)
-    return -tf.reduce_mean(policy_action_value)
+    gamma = policy.config["gamma"]
+    model = policy.model
+    n_samples = policy.config["branching_factor"]
+
+    policy_action = model.get_actions(obs)
+    next_state_dist = model.next_state_dist(obs, policy_action)
+    sampled_next_state = next_state_dist.sample((n_samples,))
+    next_state_log_prob = tf.reduce_sum(
```

Comment:
BTW... I will define a base class for Dynamics models so to make this interface standard...

Reply:
Another thing... for the same reason you stopped the gradient on `sample_next_state` when computing the `next_state_value`, you should also stop the gradient when computing `next_state_log_prob` ... otherwise the optimizer would backpropagate through the sampling process as well in addition to the backpropagation through the distribution parameters in the log_prob computation...

Reply:
> Both `sample()` and `log_prob()` methods are already implemented in the `GaussianDynamicsModel` class.
> 
> Why have you re-implemented them here?

I didn't want to use a separate forward pass through the dynamics just to compute the `log_prob`. Maybe TensorFlow takes care of optimizing this, but from my experience using PyTorch, I find it better to use the same `dist` to sample the actions and compute the `log_prob`.

Reply:
> Another thing... for the same reason you stopped the gradient on `sample_next_state` when computing the `next_state_value`, you should also stop the gradient when computing `next_state_log_prob` ... otherwise the optimizer would backpropagate through the sampling process as well in addition to the backpropagation through the distribution parameters in the log_prob computation...

Probably another effect of using PyTorch before. There, all tensors created by `sample()` from a `Distribution` have no `grad_fn`. Reparameterizable distributions have an explicit `rsample()` method  which propagates gradients.

Since I'm not sure if the same applies here, I'll call `stop_gradient` on the sample as suggested. Thanks!

Reply:
You are right it would be better not to add another forward call through the dynamics... However I think it's best to implement a `log_prob_sampled` method in the dynamics model to handle this    implementation and also to take care of the stop_gradient issue...

I will do that right away.

Reply:
I've just pushed a commit in feature/unimodal-gaussian-model with a log_prob_sampled that you can use... let me know what you think